collector/mesos_agent/mesos_agent.go

```diff
@@ -115,36 +149,109 @@ type resourceStatistics struct {
 	NetTxDropped uint64 `json:"net_tx_dropped,omitempty"`
 }
 
+func (h *MesosAgentCollector) RunPoller() {
+	ticker := time.NewTicker(h.PollPeriod * time.Second)
+
+	// Poll once immediately
+	h.pollMesosAgent()
+	for _, m := range h.transformContainerMetrics() {
+		h.MetricsChan <- m
+	}
+
+	for {
+		select {
+		case _ = <-ticker.C:
+			h.pollMesosAgent()
+			for _, m := range h.transformContainerMetrics() {
+				h.MetricsChan <- m
+			}
+		}
+	}
+}
+
+func (h *MesosAgentCollector) pollMesosAgent() {
+	now := time.Now().UTC()
+	h.timestamp = now.Unix()
+
+	host := fmt.Sprintf("%s:%d", h.NodeInfo.IPAddress, h.Port)
+
+	// always fetch/emit agent state first: downstream will use it for tagging metrics
+	maColLog.Debugf("Fetching state from DC/OS host %s", host)
+	if err := h.getAgentState(); err != nil {
+		maColLog.Errorf("Failed to get agent state from %s. Error: %s", host, err)
+	}
+
+	maColLog.Debugf("Fetching container metrics from DC/OS host %s", host)
+	if err := h.getContainerMetrics(); err != nil {
+		maColLog.Errorf("Failed to get container metrics from %s. Error: %s", host, err)
+	}
+}
+
 // getContainerMetrics queries an agent for container-level metrics, such as
 // CPU, memory, disk, and network usage.
-func (h *DCOSHost) getContainerMetrics() ([]agentContainer, error) {
-	var containers []agentContainer
+func (h *MesosAgentCollector) getContainerMetrics() error {
+	h.containerMetrics = []agentContainer{}
 	u := url.URL{
-		Scheme: "http",
-		Host:   net.JoinHostPort(h.IPAddress, strconv.Itoa(h.Port)),
+		Scheme: h.RequestProtocol,
+		Host:   net.JoinHostPort(h.NodeInfo.IPAddress, strconv.Itoa(h.Port)),
 		Path:   "/containers",
 	}
 
 	h.HTTPClient.Timeout = HTTPTIMEOUT
 
-	return containers, Fetch(h.HTTPClient, u, &containers)
+	return http_client.Fetch(h.HTTPClient, u, &h.containerMetrics)
 }
 
 // getAgentState fetches the state JSON from the Mesos agent, which contains
 // info such as framework names and IDs, the current leader, config flags,
 // container (executor) labels, and more.
-func (h *DCOSHost) getAgentState() (agentState, error) {
-	state := agentState{}
+func (h *MesosAgentCollector) getAgentState() error {
+	h.agentState = agentState{}
```

Comment:
It returns the return from Fetch() which requires you to pass reference of the struct which it unmarshal's JSON to.